src/backend/distributed/planner/distributed_planner.c

```diff
@@ -2692,43 +2798,69 @@ ErrorIfMergeHasUnsupportedTables(Query *parse, List *rangeTableList)
 			}
 			else
 			{
+				/* Usually we don't reach this exception as the Postgres parser catches it */
 				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								errmsg("MERGE command is not allowed "
-									   "on materialized view")));
+								errmsg("MERGE command is not allowed on "
+									   "relation type(relkind:%c)",
+									   rangeTableEntry->relkind)));
 			}
 		}
 
 		if (rangeTableEntry->relkind != RELKIND_RELATION &&
 			rangeTableEntry->relkind != RELKIND_PARTITIONED_TABLE)
 		{
-			ereport(ERROR,
-					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("Unexpected relation type(relkind:%c) in MERGE command",
-							rangeTableEntry->relkind)));
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("Unexpected table type(relkind:%c) in MERGE command",
+								   rangeTableEntry->relkind)));
 		}
 
 		Assert(rangeTableEntry->relid != 0);
 
-		/* Distributed tables and Reference tables are not supported yet */
-		if (IsCitusTableType(relationId, REFERENCE_TABLE) ||
-			IsCitusTableType(relationId, DISTRIBUTED_TABLE))
+		/* Reference tables are not supported yet */
+		if (IsCitusTableType(relationId, REFERENCE_TABLE))
 		{
-			ereport(ERROR,
-					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("MERGE command is not supported on "
-							"distributed/reference tables yet")));
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("MERGE command is not supported on reference "
+								   "tables yet")));
+		}
+
+		/*
+		 * For now, save all distributed tables, later (below) we will
+		 * check for supported combination(s).
+		 */
+		if (IsCitusTableType(relationId, DISTRIBUTED_TABLE))
+		{
+			distTablesList = lappend(distTablesList, rangeTableEntry);
+			continue;
 		}
 
 		/* Regular Postgres tables and Citus local tables are allowed */
 		if (!IsCitusTable(relationId) ||
 			IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
 		{
+			foundLocalTables = true;
 			continue;
 		}
 
-
 		/* Any other Citus table type missing ? */
 	}
 
-	/* All the tables are local, supported */
+	/* Ensure all tables are indeed local */
+	if (foundLocalTables)
+	{
+		if (list_length(distTablesList) != 0)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("MERGE command is not supported with "
+								   "combination of distributed/local tables yet")));
+		}
+		else
+		{
+			/* All the tables are local, supported */
+			return;
+		}
+	}
```

Comment:
```suggestion
	/* Ensure all tables are indeed local */
	if (foundLocalTables && list_length(distTablesList) == 0)
	{
		/* All the tables are local, supported */
		return;
	}
	else if (foundLocalTables && list_length(distTablesList) > 0)
	{
		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
						errmsg("MERGE command is not supported with "
								"combination of distributed/local tables yet")));
	}
```

Maybe we can get rid of 1 level of indenting as it seems easier to follow.

Reply:
Done.